### PR TITLE
Don't publish ReactAndroid to NPM

### PR DESCRIPTION
**Pull request by satya164**
The `ReactAndroid` folder contains sources for the android code. We publish the binary to NPM, so doesn't make sense to publish the source. This adds unnecessary 5MB to the download.

cc @bestander

### CODE REVIEW

**Comment by bestander** [comment]
Yeah, android folder contains the .jar with the sources.
Should be sufficient.

**Comment by bestander** [comment]
@facebook-github-bot shipit

**Comment by bestander** [comment]
We'll have feedback in 0.24.0-rc

**Comment by chirag04** [comment]
@satya164 @bestander For those who wants to compile from source on master or tinker with RN while developing, `ReactAndroid` is needed.

Can we not ship this.

**Comment by satya164** [comment]
@chirag04 For that, you'll usually maintain a forked git repo, right?

**Comment by satya164** [comment]
Ah, NPM strips this while installing from GIT repo also. @bestander is there a way we could apply this only in releases, but keep this in master?

**Comment by chirag04** [comment]
@satya164 exactly. It's a good idea to remove it only in releases if `5MB` is an overhead.

Also: There original commit by @mkonicek : https://github.com/satya164/react-native/commit/403176ae9332fb7c8beac339dd41cbb067f995fd

**Comment by bestander** [comment]
Ok, I'll stop the build

**Comment by bestander** [comment]
There is a distinction between -rc and release builds in [npm publish CI script](https://github.com/facebook/react-native/blob/master/scripts/publish-npm.js).

We could have easily done this change in the script.

But maybe we should not?
If you think about it, 5MB is just 12% of total npm download we do and 1% of compiled node_modules.
For the saved bandwidth we pay:
- inconsistency between -rc and -release builds
- breaking people who may patch React Native java code in their production systems for whatever reason

File: scripts/publish-npm.js [hunk 1]
```diff
@@ -117,6 +117,13 @@ artifacts.forEach((name) => {
   }
 });
 
+// don't publish `ReactAndroid` to npm
+const pak = JSON.parse(fs.readFileSync('package.json'));
+
+pak.files = pak.files.filter(filename => filename !== 'ReactAndroid');
+
+fs.writeFileSync('package.json', JSON.stringify(pak, null, 2));
```
Comment (bestander) [inline scripts/publish-npm.js]: JSON.stringify(packageJson, null, 2).to(`package.json`);

IF we are using shelljs, better not to mix node api with it unless necessary

**Comment by satya164** [comment]
@bestander 

> inconsistency between -rc and -release builds

Why would that happen?

> breaking people who may patch React Native java code in their production systems for whatever reason

Won't people who are patching RN usually maintain fork? They have the option of just cloning the repo instead of installing from NPM and patching it. So...

Yesterday it took me over half an hour to finish `react-native init`. Okay, my internet is a bit slow, but I'd  prefer to trim every bit of fat I can. Both dependencies, and crore.

File: scripts/publish-npm.js (hunk 1 above)
Comment (satya164) [inline scripts/publish-npm.js]: @bestander Not familiar with shelljs. Will change it :)

File: scripts/publish-npm.js [hunk 2]
```diff
@@ -125,6 +132,9 @@ if (releaseVersion.indexOf(`-rc`) === -1) {
   exec(`npm publish --tag next`);
 }
 
+// undo ignore ReactAndroid
+exec(`git checkout ReactAndroid/gradle.properties`);
```
Comment (bestander) [inline scripts/publish-npm.js]: is it intended? we have the same at line 105.

**Comment by satya164** [comment]
@bestander fixed

**Comment by bestander** [comment]
>> inconsistency between -rc and -release builds
> Why would that happen?

Disregard this, I thought you want to leave the code in `-rc` but strip in releases.

>> breaking people who may patch React Native java code in their production systems for whatever reason
> Won't people who are patching RN usually maintain fork? They have the option of just cloning the repo instead of installing from NPM and patching it. So...

For example, people want to keep using stable releases but need to patch one line in native code for some reason.
They could write an npm plugin that would `sed` the line and not worry about managing a fork which needs to be updated daily.
Not a great example but these things happen.

> Yesterday it took me over half an hour to finish react-native init. Okay, my internet is a bit slow, but I'd prefer to trim every bit of fat I can. Both dependencies, and crore.

That is the problem with npm, for some reason many people had problems yesterday.
npm CLI makes 1400 requests and downloads 40MB when doing `react-native init`, I think the problem may be in the number of requests then download size.

You would save more if we get rid of yeoman with an ad-hoc bootstrapping script. :)

**Comment by satya164** [comment]
> They could write an npm plugin that would sed the line and not worry about managing a fork which needs to be updated daily.

Can't they just clone the relevant stable branch?

> npm CLI makes 1400 requests and downloads 40MB when doing react-native init, I think the problem may be in the number of requests then download size.

Yeah, sad fact :(

> You would save more if we get rid of yeoman with an ad-hoc bootstrapping script. :)

I can try working on that.

**Comment by bestander** [comment]
@mkonicek do you have any info?
Looks like you've added ReactAndroid to npm package.

**Comment by ide** [comment]
What's the performance impact of distributing 5MB over npm? I think it could be low impact because npm sends packages as gzipped tarballs, so the Java source is sent with the same HTTP request for the rest of the react-native package. This is much better than having 100s of small packages that add up to 5MB. Also, is the 5MB statistic for gzipped or non-gzipped source code? That would make a big difference too.

**Comment by sreesharp** [comment]
> For example, people want to keep using stable releases but need to patch one line in native code for some reason.
I am doing that currently to support the binary data in web socket ( https://github.com/facebook/react-native/pull/6327)

**Comment by satya164** [comment]
@sreesharp What's your process of doing it?

If you have script which patches things under node_modules at runtime, its a pretty fragile thing, and can break your build/fail at any point of time we change the code. I don't think we can support the use case even if we wanted to. But even then, you still have the option to point to `git://github.com/facebook/react-native.git#0.22-stable` in your `package.json` instead of `^0.22.0`

@ide I think even if it's low impact, its low-effort, so no reason not to do it, unless there's a convincing use case which cannot be fulfilled if we remove it.

**Comment by ide** [comment]
I zipped up ReactAndroid (`tar -czf ReactAndroid.tar.gz ReactAndroid`) and the output file was 0.47MB. Compression of the entire npm package is probably even better by a small amount since the compression tables can optimize across the Java and JavaScript code.

This is the perf improvement of getting rid of yeoman BTW (npm 3.8.3 on Node 5.9.1, node_modules fully cached):
```
$ git clean -dfx
$ time npm install
real	1m28.028s
user	0m41.213s
sys	0m11.073s

$ git clean -dfx
Removing node_modules/
$ emacs package.json 
# Removed yeoman-environment and yeoman-generator

$ time npm install
real	1m15.933s
user	0m35.906s
sys	0m9.575s
```

**Comment by satya164** [comment]
@ide Interesting. I guess it's negligible in that case. I'll work on getting rid of yoman.

BTW is your npm install really 1 min 15 sec? I'm so jealous.

**Comment by ide** [comment]
@satya164 That's with a completely warm npm cache. And this is my ping time, which is about as long as it should take npm to return an HTTP 304 response code:
```
$ ping registry.npmjs.org
PING a.sni.fastly.net (23.235.47.162): 56 data bytes
64 bytes from 23.235.47.162: icmp_seq=0 ttl=58 time=13.586 ms
64 bytes from 23.235.47.162: icmp_seq=1 ttl=58 time=13.341 ms
64 bytes from 23.235.47.162: icmp_seq=2 ttl=58 time=13.430 ms
64 bytes from 23.235.47.162: icmp_seq=3 ttl=58 time=14.412 ms
64 bytes from 23.235.47.162: icmp_seq=4 ttl=58 time=14.907 ms
64 bytes from 23.235.47.162: icmp_seq=5 ttl=58 time=19.775 ms
64 bytes from 23.235.47.162: icmp_seq=6 ttl=58 time=13.187 ms
64 bytes from 23.235.47.162: icmp_seq=7 ttl=58 time=13.391 ms
64 bytes from 23.235.47.162: icmp_seq=8 ttl=58 time=13.301 ms
64 bytes from 23.235.47.162: icmp_seq=9 ttl=58 time=21.626 ms
64 bytes from 23.235.47.162: icmp_seq=10 ttl=58 time=13.763 ms
^C
--- a.sni.fastly.net ping statistics ---
11 packets transmitted, 11 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 13.187/14.974/21.626/2.773 ms
```

These are some more npm stats:

npm 3 with progress bar turned off:
```
real	1m23.517s
user	0m40.852s
sys	0m10.450s
```

npm 2:
```
real	1m5.661s
user	1m6.900s
sys	0m26.771s
```
So npm 2 is faster for installing, but you'll spend more time in the packager since Babel will be so slow with extra dependencies. I'd go with npm 3 at this point.

And npm 3 with cold cache (`npm cache clean`):
```
real	2m39.964s
user	0m44.619s
sys	0m12.177s
```

**Comment by sreesharp** [comment]
@satya164 I have the forked to private Github, my custom changes are there in private repo. Once in a while I merge the latest from the public repo. So its manual fix and not automated patch. My build server just uses the private repo to build ReactNative.

**Pull request closed without merging**